### PR TITLE
계산기 [STEP 1] MINT

### PR DESCRIPTION
**Pull request by mint3382**
@JoSH0318 
안녕하세요 밈! 드디어 밈의 리뷰이가 되었습니다! 
3주간 잘 부탁드립니다~

# STEP 1
### Queue의 개념
선입선출! 의 개념을 처음에 단순 삭제로 잘못 이해했습니다. 때문에 `dequeue`를 하면 맨 처음에 있는 `element`가 삭제되는 것만을 생각해서 제대로 된 결과물이 나오지 않는 일이 있었습니다. `T?`를 반환값으로 주어 애초에 `enqueue`된 것이 없거나 이미 요소들이 전부 `dequeue`되어 나간 경우 `nil`을 반환할 수 있게 변경하였습니다. 이후 `Queue`의 개념에 대해 더 알아보다가 보통 `count`라는 개수를 세어주는 기능이 있길래 연산 프로퍼티를 사용해서 추가하였습니다. `subscript`를 사용해보고 싶었는데 그렇게 하면 `CalculatorItemQueue.count` 로 사용할 수 없고 `CalculatorItemQueue[1]`로 사용해야 하기에 변경하였습니다.

### generic type의 사용
계산기 프로젝트에서 이 `queue`를 사용한다면 입력받는 숫자들과 사칙연산 기호를 넣고 순서대로 맞춰서 선입선출하며 계산해주는 것에 사용할 것이라고 생각했습니다. 그런데 계산기에서는 단순한 `Int` 타입 뿐만 아니라 10.1과 같은 `Double` 타입도 사용하고, 사칙연산 기호 같은 경우는 `String` 타입을 사용할 것이기에 `Queue`의 구현에서 여러가지 타입들 중 선택할 수 있는, 범용적인 타입이 필요하다고 생각했습니다. 때문에 `Generic type`을 사용하여 `Queue`를 구현하였습니다.

### list 구현의 의미
프로젝트 스텝 안내에 나와있는 `list 자료구조 구현`의 의미를 정확하게 이해하지 못하고 있습니다. 기존 스위프트의 `array`에서 제공하는 `removeFirst()` 에서 시간복잡도를 더 낮춘 경우들을 구현해보는 것을 말하는 걸까요? ex) `linked-list`, `stack`

### test와 private 사용
`private`을 걸어주면 `test`에서도 그 함수를 사용할 수가 없었습니다. 고민하다 결국 `test`를 전부 다 할 때까지는 `private` 키워드 없이 진행하였고 마지막에 `private`이 필요하다고 생각되는 것들에만 붙여주었습니다. `count`, `enqueue`, `dequeue`는 이 `CalculatorItemQueue`를 인스턴스해서 사용할 때 불러야 하는 기능들이라고 생각해 `private` 키워드를 달지 않았습니다. 

### TDD의 활용
활동학습에서 봤던 방법을 따라했는데 제대로 된 방법이었는지는 잘 모르겠습니다.😵 
1. 먼저 선언만 한 후 테스트 함수를 작성합니다. 
    -> 이를 돌리면 `fail`이 뜹니다. 실패한 케이스.
2. 구현하고자 하는 함수 내부를 작성합니다.
    -> `refactor`
3. 다시 테스트를 돌립니다.
    -> 성공 시 다음 함수로, 실패 시 다시 리펙토링
    
### Queue 구현 방법들 중 선택사항: Double stack
queue를 구현하는 방법들은 여러가지가 있었습니다.
1. 기본적인 array 
    _ `removeFirst()`
2. linked list
    _ 본인 이전의 정보와 다음의 정보를 알고 있는 `linked list`로 맨 처음의 `linked list`를 해제하면 됩니다.
3. Double Stack
    _ `stack` 2개를 사용해 구현합니다. 두 개의 스택용 배열을 만들어서 1번 스택에 `enqueue`의 값들을 순서대로 추가합니다. 1번 스택의 요소들을 뒤에서부터 빼면서 2번 스택에 넣습니다. 2번 스택은 1번 스택을 거꾸로 뒤집은 것 같은 모양을 하게 됩니다. 2번 스택의 제일 위에 있는 요소를 빼냅니다. 다시 2번 스택을 뒤에서부터 빼면서 1번 스택에 넣으면 1번 스택은 기존에서 첫번째 요소를 제외하고 나머지 요소들이 앞당겨진 모양이 됩니다. 
    
=> 저는 이 중에서 `Double Stack`을 사용했습니다. 1번 방법은 간편하지만 맨 처음 요소를 삭제한 후 그 뒤의 것들을 하나씩 끌어와야해서 시간복잡도에 관한 문제가 생기게 됩니다. `linked list`나 `Double Stack`은 둘 다 시간 복잡도 문제를 해결해주지만 제가 먼저 생각이 난 것은 `Double Stack`이었기에 이것으로 구현하였습니다. 이렇게 시간복잡도가 같을 때는 두 가지 중에서 어떤 기준으로 선택하는 것이 좋을까요?

### reversed 기능을 가진 함수
`Double Stack`을 구현하기 위해 `reversed()`를 두번 해주어야 합니다. 이 `reversed()` 기능은 `dequeue`에서 바로 해줘도 되지만, `test`를 할 때 거꾸로 잘 바뀌는지도 확인해보고 싶어서 따로 빼주었습니다. 이렇게 빼는 것이 오히려 구현 시 복잡하게 만드는 걸까요?

### CODE REVIEW

File: Calculator/Calculator/CalculatorItemQueue.swift [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import Foundation
+
+struct CalculatorItemQueue<T>: CalculatorItem {
+    private var firstStack: [T] = []
+    private var backwardStack: [T] = []
+    var count: Int {
+        return firstStack.count
+    }
+    
+    mutating func enqueue(_ input: T){
```
Comment (JoSH0318) [inline Calculator/Calculator/CalculatorItemQueue.swift]: `)`와 `{` 사이에 공백이 필요해보입니다. 😀

File: Calculator/Calculator/CalculatorItemQueue.swift [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+//  CalculatorItemQueue.swift
+//  Calculator
+//
+//  Created by mint on 2023/05/30.
+//
+
+import Foundation
+
+struct CalculatorItemQueue<T>: CalculatorItem {
+    private var firstStack: [T] = []
```
Comment (JoSH0318) [inline Calculator/Calculator/CalculatorItemQueue.swift]: first가 뜻하는게 무엇인가요? 🤔
해당 프로퍼티가 무엇을 뜻하는지 네이밍으로는 파악할 수 없을 것 같아요!
조금 더 명확한 네이밍이 필요해보입니다.

File: Calculator/Calculator/CalculatorItemQueue.swift [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    private var backwardStack: [T] = []
+    var count: Int {
+        return firstStack.count
+    }
+    
+    mutating func enqueue(_ input: T){
+        firstStack.append(input)
+    }
+    
+    private mutating func reversedStack(_ stack: [T]) -> [T] {
```
Comment (JoSH0318) [inline Calculator/Calculator/CalculatorItemQueue.swift]: `reversed`는 동사가 아닌 것 같아요😀

File: Calculator/Calculator/CalculatorItemQueue.swift [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        let output: [T]
+        output = stack.reversed()
+        
+        return output
```
Comment (JoSH0318) [inline Calculator/Calculator/CalculatorItemQueue.swift]: 문법이 조금 어색한 것 같아요!
`output` 선언과 동시에 할당하면 될 것 같아요!

File: Calculator/Calculator/CalculatorItemQueue.swift [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    }
+    
+    mutating func dequeue() -> T? {
+        guard !firstStack.isEmpty else {
+            return nil
+        }
+        
+        backwardStack = reversedStack(firstStack)
+        let output = backwardStack.popLast()
+        firstStack = reversedStack(backwardStack)
```
Comment (JoSH0318) [inline Calculator/Calculator/CalculatorItemQueue.swift]: 1. dequeue를 할 때마다 reversed를 두번씩하게 될 것 같아요. 
만약 10개의 element가 있고 dequeue를 10번 한다면, 20번을 reversed해줘야 할 것 같아요.
1번째 dequeue가 발생할 때, 첫 reversed를 하게된다면 `backwardStack`에는 뒤집힌 stack이 할당 될 것이고,
이 `backwardStack`을 활용해서 dequeue를 해줄 수 있을 것 같아요. 2번째 reversed에서 다시 뒤집어줄텐데 굳이 다시 뒤집는 동작이 조금 비효율적으로 느껴집니다. 이부분에 대해서 어떻게 생각하시나요? 😀

2. 이 과정을 거치면 같은 정보이지만 순서만 뒤바뀐 stack이 두개 존재할 것 같아요. 
만약 enqueue 1001번 반복됐고, dequeue를 1번 했다면 1000개의 element가 있는 두 개의 배열을 갖게 되겠네요. 
이것에 대해서 어떻게 생각하시나요?

File: Calculator/CalculatorTests/CalculatorTests.swift [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        try super.setUpWithError()
+        sut = CalculatorItemQueue()
+    }
+
+    override func tearDownWithError() throws {
+        try super.tearDownWithError()
+        sut = nil
+    }
+    
+    func test_count가_작동하는지() {
```
Comment (JoSH0318) [inline Calculator/CalculatorTests/CalculatorTests.swift]: `작동하는지`에서 민트가 생각하는 `작동`의 정의는 무엇인가요? 더 나아가서 "정상적인 작동"은 무엇을 의미하나요?
count를 통해서 어떠한 기대값을 갖고, 그 기대값을 충족시켰을 때 "정상적인 작동을 했다."라고 보는데 어떻게 생각하시나요?

File: Calculator/CalculatorTests/CalculatorTests.swift [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        sut.enqueue(2)
+        
+        //when
+        let result = sut.count
+        
+        //then
+        XCTAssertEqual(result, 2)
+    }
+    
+    func test_firstStack에_순서대로요소가추가되는지() {
```
Comment (JoSH0318) [inline Calculator/CalculatorTests/CalculatorTests.swift]: 1. 이 test는 어떤 메서드를 검증하는 test인가요? 
2. test 이름이 포괄적이라는 생각이 들어요. 때문에 test가 명확해보이지 않아요.
`test_enqueue_1_을_enqueue했을때_count_1을_반환한다`처럼 
어떤 메서드에 대한 test인지, 어떤 상황에서 어떤 결과값을 반환해야 올바른 동작을 했다고 봐야하는지 
명확하게 설명할 수 있는 네이밍을 고민해보셔야 할 것 같아요.
3. enqueue에 대한 test 없이, `순서 확인`을 하는 test까지 하고 있다는 생각이 들어요. test를 좀 더 세분화시켜야 할 것 같아요.

File: Calculator/CalculatorTests/CalculatorTests.swift [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        //given
+        let input = 10
+        let nextInput = 20
+        
+        //when
+        sut.enqueue(input)
+        sut.enqueue(nextInput)
+        
+        //then
+        XCTAssertEqual(sut.firstStack, [10, 20])
```
Comment (JoSH0318) [inline Calculator/CalculatorTests/CalculatorTests.swift]: `firstStack` 배열을 은닉화했기 때문에 오류가 뜨는 코드입니다.
`firstStack` 배열을 조회하기 위한 다른 방법을 생각하는 게 좋을 것 같아요!

File: Calculator/CalculatorTests/CalculatorTests.swift [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        sut.enqueue(3)
+        
+        //when
+        let result = sut.reversedStack(sut.firstStack)
+        
+        //then
+        XCTAssertEqual(result, [3, 2, 1])
+    }
+    
+    func test_deQueue_하면선입선출결과나오는지() {
```
Comment (JoSH0318) [inline Calculator/CalculatorTests/CalculatorTests.swift]: deQueue 오타인 것 같습니다. 😀

**Review by JoSH0318 — COMMENTED** [review]
안녕하세요 😀 민트🍵
첫번째 pr 하시느라 고생하셨습니다. 👍
요구사항에 맞춰서 깔끔하게 pr해주신 것 같아요.
다만, 아직 unit test가 처음이다보니 익숙하지 않아보여요. 😆
조금만 더 고민하시고, 익숙해진다면 좋은 코드가 될 거라 확신해요. 👍

몇가지 코멘트, 질문, 요구사항, 답변 남겼습니다. 
리뷰에 대해서 이해가 안되거나 궁금하신 사항있으시면 언제든지 DM주세요. 😀

---

### 질문
1️⃣ List 자료구조와 DoubleStackQueue의 장단점은 무엇이 있을까요? 

---

### 요구사항
1️⃣ test method을 좀 더 세분화하고, 다양한 경우와 edge case을 고려한 test 코드를 작성해주시면 좋을 것 같아요.😀 현재로썬 `특정 기능을 검증했다.`라고 보기엔 test가 부족하다는 생각이 들어요!

2️⃣  pr할 때, 오류가 있는 코드는 최대한 피해주세요!

3️⃣ MVC에 맞게 파일트리 정리를 해주셨으면 좋겠습니다.😀

4️⃣ 계산기에는 AC(All Clear)버튼이 있습니다. 때문에 Queue를 전부 지워주는 기능도 필요할 것 같아요. 이처럼 확장성을 고려한 기능을 추가적으로 구현해주시면 좋을 것 같습니다.☺️

5️⃣ commit message에 test 관련 commit을 `feat`으로 작성하셨는데, 어떠한 이유일까요? 해당 commit이 test라는 것을 파악하기 어려워보입니다. 🥹

---

### 답변
1️⃣ TDD의 활용
Q: 활동학습에서 봤던 방법을 따라했는데 제대로 된 방법이었는지는 잘 모르겠습니다.😵
먼저 선언만 한 후 테스트 함수를 작성합니다.
-> 이를 돌리면 fail이 뜹니다. 실패한 케이스.
구현하고자 하는 함수 내부를 작성합니다.
-> refactor
다시 테스트를 돌립니다.
-> 성공 시 다음 함수로, 실패 시 다시 리펙토링

> A: 활동학습을 열심히 들으셨네요! 개인적인 견해로 민트의 판단이 맞다고 봅니다. 다만, TDD도 하나의 개발 방법론이기 때문에 예외도 있을 것이고, 팀/개인별 성향에 따라 디테일은 조금씩 변할 수도 있다고 봅니다.👍 

2️⃣ Queue 구현 방법들 중 선택사항: Double stack
Q: queue를 구현하는 방법들은 여러가지가 있었습니다.
기본적인 array
_ removeFirst()
linked list
_ 본인 이전의 정보와 다음의 정보를 알고 있는 linked list로 맨 처음의 linked list를 해제하면 됩니다.
Double Stack
_ stack 2개를 사용해 구현합니다. 두 개의 스택용 배열을 만들어서 1번 스택에 enqueue의 값들을 순서대로 추가합니다. 1번 스택의 요소들을 뒤에서부터 빼면서 2번 스택에 넣습니다. 2번 스택은 1번 스택을 거꾸로 뒤집은 것 같은 모양을 하게 됩니다. 2번 스택의 제일 위에 있는 요소를 빼냅니다. 다시 2번 스택을 뒤에서부터 빼면서 1번 스택에 넣으면 1번 스택은 기존에서 첫번째 요소를 제외하고 나머지 요소들이 앞당겨진 모양이 됩니다.
=> 저는 이 중에서 Double Stack을 사용했습니다. 1번 방법은 간편하지만 맨 처음 요소를 삭제한 후 그 뒤의 것들을 하나씩 끌어와야해서 시간복잡도에 관한 문제가 생기게 됩니다. linked list나 Double Stack은 둘 다 시간 복잡도 문제를 해결해주지만 제가 먼저 생각이 난 것은 Double Stack이었기에 이것으로 구현하였습니다. 이렇게 시간복잡도가 같을 때는 두 가지 중에서 어떤 기준으로 선택하는 것이 좋을까요?

> A: Linked list와 Double Stack에 대해서 학습하면 답을 찾을 수 있을 것이라 판단됩니다.😀 두 가지 방법은 명확한 차이가 있어요! 그 부분을 찾아내신다면 질문에 대한 해답도 찾을거라 확신합니다.

3️⃣ reversed 기능을 가진 함수
Q: Double Stack을 구현하기 위해 reversed()를 두번 해주어야 합니다. 이 reversed() 기능은 dequeue에서 바로 해줘도 되지만, test를 할 때 거꾸로 잘 바뀌는지도 확인해보고 싶어서 따로 빼주었습니다. 이렇게 빼는 것이 오히려 구현 시 복잡하게 만드는 걸까요?

> A: 아하! 좋은 고민입니다. 제가 질문 드리고 싶은게 있습니다. swift에서 제공하는 instance method도 test해야할까요?  민트의 고민을 다시 정리해서 말하자면 "swift에서 제공하는 instance method도 test하고 싶다"가 되는 것 같아요! 제 개인적인 견해로 이 부분은 test가 필요없다고 봅니다.😀

File: Calculator/CalculatorTests/CalculatorTests.swift (hunk 7 above)
Comment (mint3382) [reply Calculator/CalculatorTests/CalculatorTests.swift]: 최대한 정확한 이름으로 test를 지어보려고 노력하여 수정해 보았습니다! 그런데 그러다보니 test명이 엄청나게 길어졌는데 괜찮을까요? 또 enqueue에 대한 test만 따로 하기는 힘들어서 count와 enqueue를 동시에 이루어 진다고 생각하고 count로 검증하였습니다.

File: Calculator/CalculatorTests/CalculatorTests.swift (hunk 6 above)
Comment (mint3382) [reply Calculator/CalculatorTests/CalculatorTests.swift]: 저도 그렇게 생각합니다...! 수정했습니다🥰 감사합니다~!

File: Calculator/Calculator/CalculatorItemQueue.swift (hunk 5 above)
Comment (mint3382) [reply Calculator/Calculator/CalculatorItemQueue.swift]: 이러한 부분 때문에 제가 구현한 double stack은 double stack의 장점을 살리지 못하고 있는 것 같아 변경해보았습니다. 
enqueueStack은 enqueue만 해주고, dequeueStack은 비어 있는 경우 enqueueStack을 거꾸로 뒤집어 가져온 후 기존 enqueueStack은 전부 삭제합니다. 이렇게 하면 dueueStack에서 dequeue를 queue의 정의에 맞게 할 수 있게 됩니다. 또한 매번 reversed()를 두번 할 필요도 없어서 비효율적인 동작도 제거할 수 있습니다.

**Comment by mint3382** [comment]
### linked-list와 double stack
linked-list와 double stack을 전부 구현해 보았습니다. 

linked list는 node라는 곳에 단순한 본인의 data 뿐만 아니라 다음 차례의 node의 주소값을 알고 있는 구조입니다. 끊임없이 연결되어 있는 구조이고 HEAD와 TAIL을 설정해주면 손쉽게 처음과 끝을 알 수 있고, 내용을 변경할 수 있습니다. 때문에 array에서 구현되는 removeFirst()와는 다르게 HEAD만 옮겨서 1번 자리를 가리키는 것이 바뀔 뿐 배열의 내용을 앞으로 당길 필요가 없기에 시간복잡도가 O(n)이 되는 경우가 없어서 빠릅니다. 또한 중간에 데이터A를 넣고 싶을 때 넣고 싶은 부분의 앞의 노드의 다음 노드에 대한 정보를 A에 대한 노드로 바꾸고, A 노드에서 다음 노드에 대한 정보를 기존에 넣고 싶었던 부분의 뒤의 노드로 변경해주면 됩니다. 이처럼 데이터를 중간에 삽입하는 것에 큰 장점이 있습니다. 

Double stack은 dequeueStack이 비어있는 경우 enqueueStack을 뒤집어주어야 하는 일이 있어서 시간 복잡도는 조금 더 높을 수 있지만 배열을 이용한 구현이기에 원한다면 인덱스를 통한 접근을 빠르게 할 수 있습니다. 

계산기의 경우 큐의 구현이기에 중간에 데이터를 추가해 줄 필요가 없고 linked-list를 사용한다면 enqueue를 할 때마다 매번 node를 생성하고 할당하여 TAIL과 연결짓는 작업을 해주어야 하는데 double stack은 단순히 enqueueStack에 append만 해주면 되고, dequeue를 할 때도 linked-list는 기존 HEAD에서 데이터를 저장하여 반환한 후 HEAD가 가리키고 있는 노드를 다음 노드로 변경하는 작업을 해주어야 하는데 double stack에서는 이미 dequeueStack에 있다면 단순히 배열의 마지막 요소를 꺼내기만 하는 것이기에 해야하는 작업이 더 적고 구조가 간단하다는 생각이 들어 double stack으로 구현한 것을 남겼습니다.

File: Calculator/Calculator/Model/CalculatorItemQueue.swift [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+            dequeueStack = enqueueStack.reversed()
+            enqueueStack.removeAll()
+        }
+        
+        return dequeueStack.removeLast()
+    }
+    
+    mutating func clear() {
+        enqueueStack = []
+        dequeueStack = []
```
Comment (JoSH0318) [inline Calculator/Calculator/Model/CalculatorItemQueue.swift]: 다른 곳에선 removeAll을 호출하셨는데 빈 array를 할당한 이유가 있을까요?

File: Calculator/CalculatorTests/CalculatorTests.swift [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+    }
+    
+    func test_enqueue2번시_count가_2인지() {
+        //given
+        let input = 1
+        let input2 = 2
+        
+        //when
+        sut.enqueue(input)
+        sut.enqueue(input2)
```
Comment (JoSH0318) [inline Calculator/CalculatorTests/CalculatorTests.swift]: `enqueue가 두번 실행됐다는 주어진 조건`에서 `count를 읽었을 때`를 test하는 게 좋다고 생각합니다.
때문에 given, when, then에 맞춰 수정해보시면 어떨까요?

File: Calculator/CalculatorTests/CalculatorTests.swift [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+    
+    func test_dequeue_enqueue1과2실행시_dequeue1번_결과가_1인지() {
+        //given
+        let input = 1
+        let input2 = 2
+        
+        //when
+        sut.enqueue(input)
+        sut.enqueue(input2)
+        let result = sut.dequeue()
```
Comment (JoSH0318) [inline Calculator/CalculatorTests/CalculatorTests.swift]: 이곳도 👀

File: Calculator/CalculatorTests/CalculatorTests.swift [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+    }
+    
+    func test_dequeue_enqueue는_1번인데_dequeue는_2번일때_1번과nil이_반환되는지() {
+        //given
+        let input = 10
+        
+        //when
+        sut.enqueue(input)
+        let result = sut.dequeue()
+        let result2 = sut.dequeue()
```
Comment (JoSH0318) [inline Calculator/CalculatorTests/CalculatorTests.swift]: 이곳도 👀

File: Calculator/CalculatorTests/CalculatorTests.swift [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+        let input = 1
+        let input2 = 2
+        let input3 = 3
+        
+        //when
+        sut.enqueue(input)
+        sut.enqueue(input2)
+        sut.enqueue(input3)
+        let answer = sut.dequeue()
+        let result = sut.count
```
Comment (JoSH0318) [inline Calculator/CalculatorTests/CalculatorTests.swift]: 이곳도 👀

File: Calculator/CalculatorTests/CalculatorTests.swift [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+        let input2 = 20
+        let input3 = 30
+
+        //when
+        sut.enqueue(input)
+        sut.enqueue(input2)
+        let result = sut.dequeue()
+        sut.enqueue(input3)
+        let result2 = sut.dequeue()
+        let result3 = sut.dequeue()
```
Comment (JoSH0318) [inline Calculator/CalculatorTests/CalculatorTests.swift]: 이곳도 👀

File: Calculator/CalculatorTests/CalculatorTests.swift [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+        sut.enqueue(2)
+        sut.enqueue(3)
+        let answer = sut.dequeue()
+        
+        //when
+        let result = sut.first
+        
+        //then
+        XCTAssertEqual(result, 2)
+        XCTAssertEqual(answer, 1)
```
Comment (JoSH0318) [inline Calculator/CalculatorTests/CalculatorTests.swift]: 해당 test와 관련없는 코드라고 보여집니다.😂
사용하지 않는 반환값은 _로 처리해주시면 됩니다!

File: Calculator/Calculator/Model/CalculatorItemQueue.swift [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+        guard isEmpty == false else {
+            return nil
+        }
+        
+        if dequeueStack.isEmpty {
+            dequeueStack = enqueueStack.reversed()
+            enqueueStack.removeAll()
+        }
+        
+        return dequeueStack.removeLast()
```
Comment (JoSH0318) [inline Calculator/Calculator/Model/CalculatorItemQueue.swift:52]: `removeLast`와 `popLast`의 차이점은 무엇인가요?

**Review by JoSH0318 — COMMENTED** [review]
코멘트와 질문 남겼습니다.
추가적으로 `Test Double`에 대해서 학습해보시면 좋을 것 같습니다.😀

### 요구사항
1️⃣ UML class diagram
protocol 채택한 관계를 실선으로 표기하셨어요! 이점 확인해보시고 수정하면 좋을 것 같습니다.

File: Calculator/Calculator/Model/CalculatorItemQueue.swift (hunk 17 above)
Comment (mint3382) [reply Calculator/Calculator/Model/CalculatorItemQueue.swift:52]: removeLast는 반환값이 T이고 popLast는 T?입니다. 때문에 removeLast는 반환할 값이 없으면 빌드 과정에서 문제가 생기지만 popLast는 반환할 값이 없으면 nil을 반환합니다.

File: Calculator/Calculator/Model/CalculatorItemQueue.swift (hunk 10 above)
Comment (mint3382) [reply Calculator/Calculator/Model/CalculatorItemQueue.swift]: 헤헤 removeAll() 했던 것을 까먹었어요 헤헿 수정하였습니다!

**Comment by mint3382** [comment]
### Test Double
test double은 dummy, fake, mock, stub, spy가 있습니다. 
1. dummy: test를 위한 가짜 요소를 만들어 주는 것. 자리만 채워주는 역할
2. fake: 실제 코드에서는 필요하지만 테스트시의 그 객체의 역할에는 불필요한 것들을 전부 제거하여 핵심 코드만 남겨놓은 것.
3. mock: 리턴값과 상관없이 해당 함수가 작동하였는지 하지 않았는지를 체크하는 것.
4. stub: 끊임없이 변하기에 명확한 값을 비교해줄 수 없을 때 예시용으로 고정된 값을 만들어서 비교해주는 것.
5. spy: stub + mock 예시용으로 고정된 값도 만들어주고, 해당 함수가 작동하였는지도 체크하는 것.

다만 현재 작성하고 있는 계산기의 경우 위와 같은 것들을 굳이 구현하지 않더라도 test를 작성할 수 있기에 구현 하지는 않았습니다.

File: Calculator/Calculator/Model/CalculatorItemQueue.swift [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+    }
+    var isEmpty: Bool {
+        return enqueueStack.isEmpty && dequeueStack.isEmpty
+    }
+    var first: T? {
+        if !dequeueStack.isEmpty {
+            return dequeueStack.last
+        } else {
+            return enqueueStack.first
+        }
```
Comment (JoSH0318) [inline Calculator/Calculator/Model/CalculatorItemQueue.swift]: 부정(isEmpty)의 부정(!)이라 가독성이 떨어집니다.
때문에 개인적으로 조건문에 `!`을 사용하지 않는 편입니다.
물론 사용할 수 밖에 없는 상황이라면 어쩔수 없지만, 해당 코드에서는 피할수 있는 방법이 있을 것 같아요!
